compiler/env/AOTStorageInterface.hpp

```diff
@@ -0,0 +1,38 @@
+/*******************************************************************************
+ * Copyright (c) 2022, 2022 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at http://eclipse.org/legal/epl-2.0
+ * or the Apache License, Version 2.0 which accompanies this distribution
+ * and is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License, v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception [1] and GNU General Public
+ * License, version 2 with the OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+
+#ifndef TR_AOTSTORAGEINTERFACE_INCL
+#define TR_AOTSTORAGEINTERFACE_INCL
+
+#include "env/OMRAOTStorageInterface.hpp"
+
+namespace TR{
+
+class OMR_EXTENSIBLE AOTStorageInterface : public OMR::AOTStorageInterfaceConnector
```

Comment:
Hello Mark @mstoodle, thank you for the comment.
I was thinking of the storage classes to inherit from this class and implement load and store method. This was planned by using extensible classes, "shadowing" the OMR implementation on the project levels, hence I did not declare the functions virtual. In one of our previous discussions you advised to not mix static and dynamic parallelism, unless it's the following scenario:
"don't use `virtual` if you're composing a `TR::Class` but you can use virtual for `TR::SubClass::foo()` overriding `TR::Class::foo()` when `TR::SubClass : public TR::Class`" 
As this PR is supposed to be a first contribution, I've kept the implementation minimal. I have some more functionality implemented in my project. For example, I have two subclasses that "implement" this interface, one that uses shared objects to store the data, and another that implements shared cache functionality. Indeed, they cannot coexist in my current implementation and I have to rebuild with different define guards to use each version.
I am open to implementing  or porting any changes that you suggest. If you have anything specific in mind, can you please advise? Thank you for the review again.

Reply:
> ...and I have to rebuild with different define guards to use each version.

That's what surprised me. I did not expect to have to choose only one, which is why it struck me as virtual polymorphism (TR::SubClass::foo() where TR::SubClass : public TR::Class) than static polymorphism.

I think those two implementations should probably be separate objects and you just instantiate whichever one you need to use. The rest of the infrastructure should operate against the "interface" base class using virtual functions overridden by each implementation class.

Reply:
Hello Mark @mstoodle , Is there a way that I could keep ``AOTStorage`` extensible? The design avoiding `virtual` was inspired by the rest of the OMR compiler codebase, that seemed to rely on using Extensible Classes technology. Following your request, I have tried making the class `virtual` but it turned out to be not such a trivial task in my projects ( two people who were working on the different `storage` options have already left the group, and refactoring someone else's code from using static polymorphism into dynamic is a bit difficult). The fact that two people worked in parallel is the reason why the project needs to be rebuilt to facilitate different storage: we needed the projects to coexist and people not merging so often.

In my idea, the project working with OMR AOT should a extensible class inheriting from `OMRAOTStorageConnector` and implementing `loadEntry` and `allocateEntry`, as we've discussed in previous rounds of review. If these methods are not implemented, OMR AOT will assert, as `AOTLoadStoreDriver` has some logic working already with `TR::AOTStorage` and there's a `TR_UNIMPLEMENTED()`, so people will be directed to write that code. In these same methods, the developers may also decide which method of storage to use: could be shared cache, or ELF objects, or both really, as any class can inherit from `AOTStorage`, that's why I called it `AOTStorageInterface` before, but in fact, it's more like `AOTStorageAbstractStaticPolymorphicClass`, as `AOTLoadStoreDriver` needs an object to be instantiated to have a more complete `storeEntry` functionality (designing that with pure virtual class would require us more back and forth on determining the outlook). I have  removed the "Interface", in attempt for class name to be less confusing. 

Can you please have a look and say if this reasoning is sufficient to let the AOTStorage object into the codebase as an extensible class, or do you believe it should be rewritten as virtual interface?

Reply:
Everything you've said is an argument for polymorpishm, but not for static polymorphism (which is what extensible classes are).

>  the developers may also decide which method of storage to use: could be shared cache, or ELF objects, or both really, as any class can inherit from AOTStorage

Extensible classes will force those developers to choose one at build time. i.e. you can build a compiler that uses a shared cache, or another compiler that uses an ELF object. But if you only use extensible classes without virtual hierarchy, you cannot have both in the same compiler. I don't think you should limit it in this way.

Extensible classes are not an *alternative* to virtual polymorphism. Extensible classes are a way to compose a class from parts provided by different compiler projects (i.e. OMR provides some base function which then gets specialized by a particular compiler (e.g. Java) and for the target platform (e.g. this is a compiler for x86, so don't bother including all the code for POWER, Z, ARM, etc.). Once built, you have a Java compiler for x86. if you want something other than Java or to target another platform, you have to build a different compiler.

How AOT code is stored *could* be made an option like that, but I think virtual polymorphism is the right approach here. Why can't the base class become just a regular class that defines abstract virtual `loadEntry`, `storeEntry`, and `allocateEntry` functions? Even just adding "virtual" to those three functions and then having your subclasses override those virtual functions without themselves being extensible would probably be enough to satisfy me (though I would still want you to have a compelling explanation and documentation for why the interface needs to be extensible). Then the different subclasses for `AOTStorage` can override those base functions as your currently extensible classes must. If they don't provide implementations, then it will be impossible to link the compiler together (because it won't be possible to initialize the subclass VFTs without implementations for those functions). I'm not sure what "designing that with pure virtual class would require us more back and forth on determining the outlook" refers to. 

Your argument about the difficulty of refactoring it is not compelling. If sub projects already find this code hard to manage, then accepting the base classes into the project is just asking the project to take on responsibility for a difficult "API". If that's the case, then I'm not sure the project should be keen to take on this additional burden...That's one of the reasons I've been pushing back on a few aspects of this design...

Reply:
Hello Mark @mstoodle , thank you for your comment. I have refactored the code, hopefully to your liking. Thank you for your feedback, I have removed the connectors and `OMR_EXTENSIBLE` notions, and made the methods `virtual`. It is a bit unclear to me if you would still be interested to have the interface promoted to TR in a separate file, can you please comment on that?

Regarding the difficulty of the change - it was a comment based on the fact that there was two years of development from the original proposal in two projects (Shared Cache and ELF), by the people who already left my team. After learning that legacy code, the change was not so difficult.

Thank you for your comment again, I will appreciate further feedback if there is any!